Reloads/resizes heatmap on map resize. Fixes #605

ahockersten opened the pull request:
fixes #605 

- Also fixes similar issues for player ward placement data

ahockersten commented:
I haven't tested this on mobile yet, because I expect to change it based on feedback. It works on desktop though.

ahockersten commented:
Before anyone asks about it: I decided to copy the contents of resize.js from some blog, because there really wasn't any cleaner way that I could find. I found lots of half-baked libs on github for solving this issue. Some of them were even based on this code, but none that seemed to be available in a finished form in NPM, and were under active maintenance. In the end, this seemed like the least bad solution.

howardchung commented on public/js/resize.js:
```diff
@@ -0,0 +1,65 @@
+  element.__resizeListeners__.push(fn);
+};
+
+module.exports.removeResizeListener = function(element, fn){
```
note that I'm not using webpack in the current version of `sql` to avoid build issues.  I'm just making each of these files a window.functionName = function(){}.

howardchung replied:
It should work to just replace module.exports.functionName with window.functionName.

On the webpack side you can just do require() and in base_layout.jade you can just write a script tag that includes this file.  Check out the current code for examples.  The idea is to be able to toggle between webpack and non-webpack easily.

ahockersten replied:
I don't think I understand your comment. I can't replace module.exports.functionName() with window.functionName() and still call window.resize.functionName() the way I currently do? Do you want me to rebase the pull request on the sql branch instead, where such a rewrite makes sense?

howardchung replied:
You could have resize JS assign an object to window with those functions.

In general yes, you should do feature development on sql branch to avoid
merge conflicts later.
On Oct 9, 2015 1:11 PM, "Anders Höckersten" <notifications@github.com>
wrote:

> In public/js/resize.js
> <https://github.com/yasp-dota/yasp/pull/710#discussion_r41672863>:
>
> > +    else {
> > +      if (getComputedStyle(element).position == 'static') element.style.position = 'relative';
> > +      var obj = element.__resizeTrigger__ = document.createElement('object');
> > +      obj.setAttribute('style', 'display: block; position: absolute; top: 0; left: 0; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;');
> > +      obj.__resizeElement__ = element;
> > +      obj.onload = objectLoad;
> > +      obj.type = 'text/html';
> > +      if (isIE) element.appendChild(obj);
> > +      obj.data = 'about:blank';
> > +      if (!isIE) element.appendChild(obj);
> > +    }
> > +  }
> > +  element.__resizeListeners__.push(fn);
> > +};
> > +
> > +module.exports.removeResizeListener = function(element, fn){
>
> I don't think I understand your comment. I can't replace
> module.exports.functionName() with window.functionName() and still call
> window.resize.functionName() the way I currently do? Do you want me to
> rebase the pull request on the sql branch instead, where such a rewrite
> makes sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yasp-dota/yasp/pull/710/files#r41672863>.
>

ahockersten replied:
I'll rebase this once sql gets merged to master then otherwise I'll probably have to rebase it again once it gets merged anyway.

howardchung replied:
Okay.  Thanks for the work, I don't think @albertcui or I would have gotten to this.

You could pull request into `sql` branch if you'd like earlier integration.

ahockersten replied:
I think I'll wait actually. In my opinion, the less work going in on the sql branch the better and this is hardly a critical fix.

albertcui commented:
Just to be clear, we're going to merge this into sql instead of master?

howardchung commented:
I think @ahockersten was planning to wait until after the sql->master merge, but I think it also works to have it merged into sql directly.

ahockersten commented:
As I said above, the less stuff unrelated to sql going in on the sql branch the better, in my opinion. Branches should be kept as small and topical as possible. There will likely be regressions from the merge of the sql branch (yes, I'm a pessimist when it comes to code :)), and there will be enough trouble chasing down those without having to look at totally unrelated things.

howardchung commented:
This is fully client-side, so I think it's probably okay.  It's up to you :)

There will almost certainly be some issues, whether it's bugs we can't find in dev or just trying to migrate 75 million matches. :P

howardchung commented:
@ahockersten if you want to rebase this go ahead.  However I've gotten rid of webpack so your include of resize.js will need to go in the base_layout.jade

ahockersten commented:
I've updated resizing and tested it on all heatmaps (including the new "performances").

howardchung commented on deprecated/match_positions.jade:
```diff
@@ -27,20 +27,45 @@ block match_content
-      container: document.getElementById('map'),
-      radius: 15*(width/600)
-    });
+    (function() {
```
I wouldn't bother updating this since it's deprecated

howardchung commented on views/match/match_performances.jade:
```diff
@@ -80,20 +80,51 @@ block match_content
                   
 append footer_assets
   script.
+    (function() {
```
should probably refactor this since it's used twice (performances/vision)

howardchung commented on views/player/player_wardmap.jade:
```diff
@@ -3,28 +3,75 @@ extend player
-    heatmap2.setData(posData[0]["sen"]);
-          
\ No newline at end of file
+    (function() {
```
same refactoring?  This one's a bit different since two maps but should be doable

albertcui commented:
Looks good for merge after duplicated code is refactored.

howardchung commented on views/match/match_performances.jade:
```diff
@@ -80,20 +80,51 @@ block match_content
+        if (heatmap) {
+          $(heatmap._renderer.canvas).remove();
+        }
+        heatmap = h337.create({
```
also duplication here in numbers within the function and outside.

ahockersten commented:
Refactoring done.

ahockersten commented:
I moved stuff around a bit more than requested. Since adjustHeatmapData was only used by our heatmaps I moved it inside resize.js. I also made it non-destructive so I could reuse the posData in an easier way.

howardchung commented:
Push the commit?

On Tue, Nov 10, 2015 at 11:21 AM, Anders Höckersten <
notifications@github.com> wrote:

> I moved stuff around a bit more than requested. Since adjustHeatmapData
> was only used by our heatmaps I moved it inside resize.js. I also made it
> non-destructive so I could reuse the posData in an easier way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yasp-dota/yasp/pull/710#issuecomment-155537579>.
>

ahockersten commented:
Finally pushed, after resolving some conflicts with current master

howardchung commented:
Awesome, thanks. @albertcui please review

albertcui commented:
Thanks!

albertcui commented:
Deployed.
